Add `Arel.safe` and `Arel::SafeValue` to handle quoting edge cases

kmcphillips opened the pull request:
## Problem

Fixes #44312

In some cases, SQL sanitation adds quotes around values in ways that are incorrect or can produce invalid SQL. This behaviour is mostly desired because we wish to be safe by default, and assume passed in values are unsafe. There is however no built in mechanism to bypass this in cases where the values are known to be safe.

ex:
```ruby
ActiveRecord::Base.sanitize_sql(["LIMIT :offset, :per_page", { offset: 12, per_page: 32 }])
> "LIMIT '12', '32'"
```


## Solution

Add an `Arel::SafeValue` class which simply tells the query builder and sanitization to pass through the value without quoting. Add a helper `Arel.safe()` for brevity and discoverability.

ex:
```ruby
ActiveRecord::Base.sanitize_sql(["LIMIT :offset, :per_page", { offset: Arel.safe(12), per_page: Arel.safe(32) }])
> "LIMIT 12, 32"
```

In trying to write some failing test cases I found that most of [ActiveRecord::Sanitization](https://github.com/rails/rails/blob/7ac90d9ad75c50129850572ed6c7dcd42fecbf3b/activerecord/lib/active_record/sanitization.rb) is untested. [Only a subset of it has test coverage](https://github.com/rails/rails/blob/7ac90d9ad75c50129850572ed6c7dcd42fecbf3b/activerecord/test/cases/sanitize_test.rb), and much of that coverage is agnostic to the underlying connection adapter. This PR pulls in the tests from and closes #44429 (@rbgrouleff).

It is difficult to write good clear examples because *the behaviour of the sanitization depends on the underlying adapter that is currently selected*. That was not clearly reflected in the docs or in tests, but is now.

@byroot @casperisfine as the regression was introduced in #42440.

kmcphillips committed:
Add test coverage and doc updates for ActiveRecord::Sanitization across MySQL/SQLite3/PostgreSQL adapters

kmcphillips commented on activerecord/test/cases/adapters/mysql2/sanitize_test.rb:
```diff
@@ -0,0 +1,98 @@
+
+  def test_sanitize_sql_for_order
+    actual = Post.sanitize_sql_for_order([Arel.sql("field(id, ?)"), [1, 3, 2]])
+    expected = "field(id, '1','3','2')" # this is invalid SQL but expected behaviour
```
This is not ideal. But it _is_ what is generated in the example in the docs.

simi replied:
:thinking: For example `Post.sanitize_sql_for_order([Arel.sql("COALESCE(id, ?)"), [1, 3, 2]])` seems valid SQL fragment (tested with PostgreSQL). Why is that considered invalid SQL?

```ruby
irb(main):005:0> Post.order(Post.sanitize_sql_for_order([Arel.sql("COALESCE(id, ?)"), [1, 3, 2]]))
  Post Load (0.3ms)  SELECT "posts".* FROM "posts" ORDER BY COALESCE(id, 1,3,2)
```

kmcphillips replied:
You are right, this is a bad example.

The original described bug was:
```ruby
ActiveRecord::Base.sanitize_sql(["LIMIT :offset, :per_page", { offset: 12, per_page: 32 }])
```

Expected to produce `"LIMIT 12, 32"` but produced `"LIMIT '12', '32'"`.

But **specifically in the MySQL adapter**. 
```sql
mysql> SELECT * FROM examples LIMIT '1', '2';
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ''1', '2'' at line 1
```

What has been hard about writing good docs and tests is that the output here differs on the adapter implementation, because how we handle quoting depends on the adapter selected in the current context.

I can update this test and comment for accuracy and clarity.

kmcphillips replied:
Yes. I can do that. But also I was trying to assert what we have in the docs matches actual behaviour. Since we don't specify in the docs that the examples come from a specific adapter. At least now I updated them to explain that behaviour depends on the DB and adapter in use.

kmcphillips commented:
I do not believe so, as `.sql` also follows the same case statements for each adapter in quoting `?` values:

```
> ActiveRecord::Base.sanitize_sql(["LIMIT :offset, :per_page", { offset: Arel.sql(12), per_page: Arel.sql(32) }])
/home/kevin/source/rails/activerecord/lib/arel.rb:40:in `initialize': no implicit conversion of Integer into String (TypeError)

> ActiveRecord::Base.sanitize_sql(["LIMIT :offset, :per_page", { offset: Arel.sql(12.to_s), per_page: Arel.sql(32.to_s) }])
=> "LIMIT '12', '32'"
```

kmcphillips commented:
The reason is that the case statements [here](https://github.com/rails/rails/blob/33e371070b06bea5edfdab7c2af518a7c968c832/activerecord/lib/active_record/connection_adapters/mysql/quoting.rb#L10-L23) and [here](https://github.com/rails/rails/blob/33e371070b06bea5edfdab7c2af518a7c968c832/activerecord/lib/active_record/connection_adapters/abstract/quoting.rb#L12-L26) are designed to not allow unquoted values through. We need to add some case, a wrapper class here, to prevent quoting from happening.

matthewd commented:
> We need to add some case, a wrapper class here, to prevent quoting from happening

Right... but that's what `Arel.sql` is. It sounds like we should be adding handling for those values in these places, rather than inventing another wrapper class?

kmcphillips committed:
Update Arel.sql() and alias to Arel.safe() in order to bypass quoting and pass in known safe values.

kmcphillips commented:
@matthewd I think you are suggesting something like this? https://github.com/rails/rails/commit/f55c34f7a55f1aae48e42f283758644e1ae5683c

It may change the behaviour of `Arel.sql` slightly. And it means the `Arel::Nodes::SqlLiteral` now is referenced inside the abstract adapter. But maybe that's fine.

casperisfine commented:
> Can we use existing Arel.sql for this purpose?

I'm not sure it's the same thing. `Arel.sql` is a SQL fragment. The need here is for a value that still need to be serialized as SQL, just not escaped.

kmcphillips commented:
And you end up calling it at a different level of abstraction in the same statement, which I don't love:

```ruby
([Arel.sql("field(id, ?)"), [Arel.sql(1), Arel.sql(3), Arel.sql(2)]])
```

gmcgibbon commented:
I think the concept of `Arel.sql` and `Arel.safe` are two similar but separate concerns assuming `Arel.sql` is only supposed to be used for fragments of raw SQL. Maybe we can just add an alias?

kmcphillips commented:
Ok, this is green again.

I've taken @gmcgibbon 's suggested approach and aliased `.sql` to `.safe`. I've added a bunch of tests and updated docs to cover this.

Good?

byroot reviewed (approved):
One minor concerns, but looks good to me overall and should allow to close a couple long standing issues, so ❤️ .

Let's wait for @matthewd though.

kmcphillips commented on activerecord/lib/arel.rb:
```diff
@@ -28,28 +28,35 @@
+    # This method should not be used with unsafe values such as request
+    # parameters or model attributes.
+    def sql(raw)
+      Arel::Nodes::SqlLiteral.new(raw.to_s)
```
I share this reservation.

It's why I had two different implementations originally. In that the `Arel::SafeValue` returned the raw value, not the stringified version. Here because `class SqlLiteral < String` you get a raise:

```ruby
> Arel::Nodes::SqlLiteral.new(1)
`initialize': no implicit conversion of Integer into String (TypeError)
```

So a middle of the road compromise would be:

```ruby
    def sql(raw)
      Arel::Nodes::SqlLiteral.new(raw)
    end

    def safe(raw)
      Arel::Nodes::SqlLiteral.new(raw.to_s)
    end
```

But again that doesn't _really_ solve it.

This all originated from `Rational` not behaving as expected for `to_s`:

```ruby
> Rational(3, 2).to_s
"3/2"
```

So back to advocating for having a separate `Arel::SafeValue` class originally proposed in c717720ab25139ae927d6dd2c8df795924cd5b52

matthewd replied:
Am I just confused? My expectation is that the result of `quote` is going to end up in an `#{}` interpolation in the middle of something that's definitely a string (of SQL).

That's also why I don't like the idea of the `safe` alias: SQL is SQL, and if the value you're passing isn't already a SQL string (or, newly introduced, a non-string that will `to_s` into something acceptably SQL-shaped), the caller is going to have to fix that in advance, regardless of how untouched we leave it.

@kmcphillips  can you help me understand how that original proposal helps Rational?

matthewd replied:
The one danger I do see here is that we might end up encouraging people to do `Arel.sql(params[:limit])`, which they "know" is an integer, and we end up back in easily-written-injections land.

Considering that (and incidentally rejecting the idea [which hasn't yet come up, but feels like it's nearby] that we have need of a generic mechanism for "I want to bypass AR touching this value, but hand it untouched to the DB driver"), I'm now wondering if what we want is instead a more targeted solution to the original problem [that we've introduced edge-case-troublesome quoting for numerics], and have an `Arel.number` or similar, which will only accept numeric values. 🤔

kmcphillips replied:
> how that original proposal helps Rational?

I'm not explaining myself clearly, and I'm mixing concepts here. I'm sorry. What I mean is that originally quoting _broke_ rational in that it added a `.to_s`. This breaks in the same way with the `.to_s`. See #44404.

I like this direction @matthewd. And you articulate better why the generic alias and and treatment of safe here I don't think is very good.

Let me take yet another stab at it, but this time with an exception simply for numerics.
